### ISSUE TITLE
Matching rule for non-empty map or a field which is not null

### ISSUE THREAD

**Issue by QMG-kazala** [DiUS/pact-jvm#727]
My problem is I have a field which is a list (an array) of maps. Each map can have elements of different types (strings or sub-maps), eg.

```
"validatedAnswers": [
    {
      "type": "typeA",
      "answers": {
        "favourite_colour": "Blue",
        "correspondence_address": {
            "line_1": "Main St",
            "postcode": "1A 2BC",
            "town": "London"
        }
      }
    },
    {
      "type": "typeB",
      "answers": {
        "first_name": "Firstname",
        "last_name": "Lastname",
      }
    }
  ]
```
and we're only interested in some of those answers.

What we really need is this: https://github.com/pact-foundation/pact-specification/issues/38, but it's planned for v.4. In the meantime we're trying a different approach.  What I'm attempting to do now is to specify that each element of the list is a  non-empty map. Another approach is to specify that each element of the list is not null. Can any of this be done using Groovy DSL?

This: 
```
validatedAnswers minLike(1) {
     type string()
     answers {
     }
}
```
doesn't work because it mean `answers` is expected to be empty ("Expected an empty Map but received Map( [...] )", see also https://github.com/DiUS/pact-jvm/issues/298).

So what I would like to do is something like this:
```
.withBody {
    validatedAnswers minLike(1) {
         type string()
         answers Matchers.map()
    }
}
```
or:
```
validatedAnswers minLike(1) {
     type string()
     answers {
             keyLike 'title', notNull()
     }
}
```
or:
```
validatedAnswers minLike(1) {
     type string()
     answers notNull()
}
```
Can it be done?

**Comment by QMG-kazala** [comment]
Anyone? Anything?
Is there a chance anything like notNull() matcher will be added in the future?

**Comment by QMG-kazala** [comment]
Oh yes!!!! Having predicate-based matchers would solve so many problems! It would be great :D 
👍 
Does it mean it's planned for version 4?

**Comment by uglyog** [comment]
Yes it will have to be a V4 change because it will require a change to the pact file format as well as how the matchers are applied.

**Comment by alexfdz** [comment]
@uglyog will this be included in V4?

**Comment by uglyog** [comment]
@alexfdz we are going to look at having sets of matching rules applied to different objects in lists. https://github.com/pact-foundation/pact-specification/issues/38 will track this.
